pkg/sysregistriesv2/system_registries_v2.go

```diff
@@ -517,12 +568,66 @@ func FindRegistry(ctx *types.SystemContext, ref string) (*Registry, error) {
 	return nil, nil
 }
 
-// tomlConfig loads the unmarshals the configuration at the specified path.
-// Note that the tomlConfig's fields will be overridden if they are set in
-// specified path.  This behavior is necessary for registries.conf.d stances
-// behavior.
-func (t *tomlConfig) loadConfig(path string) error {
+// tomlConfig loads and unmarshals the configuration at the specified path. Note
+// that v1 configs are translated into v2 and are cleared.  Use forceV2 if the
+// config must in the v2 format.
+//
+// Note that specified fields in path will replace already set fields in the
+// tomlConfig.  Only the [[registry]] tables are merged by prefix.
+func (c *tomlConfig) loadConfig(path string, forceV2 bool) error {
 	logrus.Debugf("Loading registries.conf %q", path)
-	_, err := toml.DecodeFile(path, t)
-	return err
+
+	// Save the registries before decoding the file where they could be lost.
+	// We merge them later again.
+	registryMap := make(map[string]Registry)
+	for i := range c.Registries {
+		registryMap[c.Registries[i].Prefix] = c.Registries[i]
+	}
+
+	// Load the tomlConfig. Note that `DecodeFile` will overwrite set fields.
+	c.Registries = nil // important to clear the memory to prevent us from overlapping fields
+	_, err := toml.DecodeFile(path, c)
+	if err != nil {
+		return err
+	}
+
+	if c.V1RegistriesConf.Nonempty() {
+		// Enforce the v2 format if requested.
+		if forceV2 {
+			return &InvalidRegistries{
+				s: fmt.Sprintf("registries.conf %q must be in format v2 but is in v1", path),
+			}
+		}
+
+		// Convert a v1 config into a v2 config.
+		if c.V2RegistriesConf.Nonempty() {
+			return &InvalidRegistries{s: "mixing sysregistry v1/v2 is not supported"}
+		}
+		v2, err := c.V1RegistriesConf.ConvertToV2()
+		if err != nil {
+			return err
+		}
+		c.V1RegistriesConf = V1RegistriesConf{}
+		c.V2RegistriesConf = *v2
+	}
+
+	// Post process registries, set the correct prefixes, sanity checks, etc.
+	if err := c.postProcess(); err != nil {
+		return err
+	}
+
+	// Merge the freshly loaded registries.
+	for i := range c.Registries {
+		registryMap[c.Registries[i].Prefix] = c.Registries[i]
+	}
+
+	// Clear the config's registries and append the merged set. The order will
+	// now change (go map iteration) but we never committed to that.
```

Comment:
This is technically true, but I wonder… The callers of `GetRegistries`/`TryUpdatingCache` are going to use the list as a whole in _some_ way (if they are looking for a single entry, they should use `FindRegistry`, after all).

E.g. CRI-O logs the configuration (the return value of `TryUpdatingCache`) on reload; it seems desirable for that log line not to change gratuitously.

Reply:
That's a very thoughtful point! We now enforce an order by sorting the registries by Prefix.